File: protocol-generator/elixir/lib/entity_core/store.ex
defmodule EntityCore.Store do
  @moduledoc """
  Storage — the two layers of §1.7, owned by one GenServer process (the BEAM
  idiom for shared mutable state; all access is serialized through it, so
  concurrent per-request dispatch (§4.8) and the atomic CAS put are race-free
  without locks):

      Content Store: hash → entity   (immutable, content-addressed, dedup)
      Entity Tree:   path → hash      (mutable location index)

  Paths are the canonical absolute form `/{peer_id}/rest` (§1.4); the peer
  canonicalizes before calling in.

  ## Emit pathway (§6.10 / v7.74 §6.13(c))

  Tree/content writes produce events delivered synchronously-inline (§9.4) to
  registered consumers. The hook is LIVE even with zero consumers (events are
  produced and discarded) so a future extension can register a consumer without
  the peer being rebuilt — the §6.13(c) MUST. A core-only peer registers zero.
  `event_type` derives from the null-`new_hash` rule alone: a bind to a
  `system/deletion-marker` fires `"modified"` (it has a new_hash), never
  `"deleted"`.
  """

  use GenServer

  alias EntityCore.Entity

  @type event :: %{event_type: String.t(), path: String.t(), new_hash: binary() | nil, previous_hash: binary() | nil}

  defstruct content: %{}, tree: %{}, content_consumers: [], tree_consumers: []

  # ── lifecycle ───────────────────────────────────────────────────────────

  @spec start_link(keyword()) :: GenServer.on_start()
  def start_link(opts \\ []), do: GenServer.start_link(__MODULE__, :ok, opts)

  @impl true
  def init(:ok), do: {:ok, %__MODULE__{}}

  # ── consumer registration (§6.10) ────────────────────────────────────────

  @spec register_content_consumer(GenServer.server(), (map() -> any())) :: :ok
  def register_content_consumer(s, fun), do: GenServer.call(s, {:register_content_consumer, fun})

  @spec register_tree_consumer(GenServer.server(), (event() -> any())) :: :ok
  def register_tree_consumer(s, fun), do: GenServer.call(s, {:register_tree_consumer, fun})

  # ── content store ─────────────────────────────────────────────────────────

  @spec put_entity(GenServer.server(), Entity.t()) :: :ok
  def put_entity(s, %Entity{} = e), do: GenServer.call(s, {:put_entity, e})

  @spec get_by_hash(GenServer.server(), binary()) :: Entity.t() | nil
  def get_by_hash(s, h), do: GenServer.call(s, {:get_by_hash, h})

  # ── entity tree ─────────────────────────────────────────────────────────

  @spec bind(GenServer.server(), String.t(), Entity.t()) :: :ok
  def bind(s, path, %Entity{} = e), do: GenServer.call(s, {:bind, path, e})

  @doc """
  Atomic compare-and-set bind (§3.9): `expected` is `:create_only` (path must be
  unbound), `{:match, hash}` (current binding must equal `hash`), or `:any`.
  Returns `:ok` or `:mismatch`. Atomic because it executes in one GenServer call.
  """
  @spec bind_cas(GenServer.server(), String.t(), Entity.t(), :create_only | {:match, binary()} | :any) ::
          :ok | :mismatch
  def bind_cas(s, path, %Entity{} = e, expected), do: GenServer.call(s, {:bind_cas, path, e, expected})

  @spec unbind(GenServer.server(), String.t()) :: :ok
  def unbind(s, path), do: GenServer.call(s, {:unbind, path})

  @spec hash_at(GenServer.server(), String.t()) :: binary() | nil
  def hash_at(s, path), do: GenServer.call(s, {:hash_at, path})

  @spec get_at(GenServer.server(), String.t()) :: Entity.t() | nil
  def get_at(s, path), do: GenServer.call(s, {:get_at, path})

  @doc """
  One-level listing under `prefix`: `[{segment, hash | nil, has_children?}]`,
  sorted by segment (§3.9 / §1.7). A bound child contributes a hash; a path that
  is also a prefix of deeper paths contributes `has_children?`.
  """
  @spec listing(GenServer.server(), String.t()) :: [{String.t(), binary() | nil, boolean()}]
  def listing(s, prefix), do: GenServer.call(s, {:listing, prefix})

  # ── server ────────────────────────────────────────────────────────────────

  @impl true
  def handle_call({:register_content_consumer, fun}, _from, st),
    do: {:reply, :ok, %{st | content_consumers: [fun | st.content_consumers]}}

  def handle_call({:register_tree_consumer, fun}, _from, st),
    do: {:reply, :ok, %{st | tree_consumers: [fun | st.tree_consumers]}}

  def handle_call({:put_entity, e}, _from, st), do: {:reply, :ok, do_put(st, e)}

  def handle_call({:get_by_hash, h}, _from, st), do: {:reply, Map.get(st.content, h), st}

  def handle_call({:bind, path, e}, _from, st), do: {:reply, :ok, do_bind(st, path, e)}

  def handle_call({:bind_cas, path, e, expected}, _from, st) do
    current = Map.get(st.tree, path)

    ok? =
      case expected do
        :any -> true
        :create_only -> current == nil
        {:match, h} -> current == h
      end

    if ok?, do: {:reply, :ok, do_bind(st, path, e)}, else: {:reply, :mismatch, st}
  end

  def handle_call({:unbind, path}, _from, st) do
    previous = Map.get(st.tree, path)
    st = %{st | tree: Map.delete(st.tree, path)}

    if previous != nil do
      emit_tree(st, %{event_type: "deleted", path: path, new_hash: nil, previous_hash: previous})
    end

    {:reply, :ok, st}
  end

  def handle_call({:hash_at, path}, _from, st), do: {:reply, Map.get(st.tree, path), st}

  def handle_call({:get_at, path}, _from, st) do
    entity = with h when h != nil <- Map.get(st.tree, path), do: Map.get(st.content, h)
    {:reply, entity || nil, st}
  end

  def handle_call({:listing, prefix}, _from, st), do: {:reply, do_listing(st, prefix), st}

  # ── internals ─────────────────────────────────────────────────────────────

  # §6.10 Store step: a content-store event fires only when the entity is new.
  defp do_put(st, %Entity{hash: h} = e) do
    if Map.has_key?(st.content, h) do
      st
    else
      st = %{st | content: Map.put(st.content, h, e)}
      Enum.each(st.content_consumers, fn f -> f.(%{hash: h, entity: e}) end)
      st
    end
  end

  # §6.10 Bind step: bind runs Store then Bind; a tree-change event fires only
  # when the binding at the path changes (no event on a re-bind to current hash).
  defp do_bind(st, path, %Entity{hash: h} = e) do
    st = do_put(st, e)
    previous = Map.get(st.tree, path)
    changed = previous != h
    st = %{st | tree: Map.put(st.tree, path, h)}

    if changed do
      emit_tree(st, %{
        event_type: event_type(previous, h),
        path: path,
        new_hash: h,
        previous_hash: previous
      })
    end

    st
  end

  defp emit_tree(st, event), do: Enum.each(st.tree_consumers, fn f -> f.(event) end)

  defp event_type(nil, _new), do: "created"
  defp event_type(_prev, nil), do: "deleted"
  defp event_type(_prev, _new), do: "modified"

  defp do_listing(st, prefix) do
    prefix = if String.ends_with?(prefix, "/"), do: prefix, else: prefix <> "/"
    plen = byte_size(prefix)

    st.tree
    |> Enum.reduce(%{}, fn {path, hash}, acc ->
      if byte_size(path) > plen and String.starts_with?(path, prefix) do
        rest = binary_part(path, plen, byte_size(path) - plen)

        case :binary.match(rest, "/") do
          :nomatch ->
            # direct child, bound
            note(acc, rest, hash, false)

          {i, _} ->
            # deeper child path
            note(acc, binary_part(rest, 0, i), nil, true)
        end
      else
        acc
      end
    end)
    |> Enum.map(fn {seg, {hash, children}} -> {seg, hash, children} end)
    |> Enum.sort_by(fn {seg, _, _} -> seg end)
  end

  defp note(acc, seg, hash_opt, deeper) do
    Map.update(acc, seg, {hash_opt, deeper}, fn {h, c} ->
      {if(hash_opt != nil, do: hash_opt, else: h), c or deeper}
    end)
  end
end

File: protocol-generator/ruby/lib/entity_core/store.rb
# frozen_string_literal: true

require_relative "entity"

module EntityCore
  # Storage (foundation, §1.7): the two layers.
  #
  #   Content Store: hash → entity   (immutable, content-addressed, dedup)
  #   Entity Tree:   path → hash     (mutable location index)
  #
  # In-memory minimal impl. Paths are the canonical absolute form
  # +/{peer_id}/rest+ (§1.4); the peer canonicalizes before calling in. The
  # content store is keyed by the LOWERCASE-HEX content_hash (A-CL-009 trap), so
  # an ASCII-8BIT byte hash works as a String map key.
  #
  # == §4.8 data-race safety (the Ruby concurrency correctness point)
  #
  # The transport dispatches concurrent inbound EXECUTEs on separate threads,
  # each of which may read-then-write the store (§3.9 CAS put, §6.13(a) register
  # multi-bind). Under MRI the GVL serializes Ruby BYTECODE, but it is RELEASED
  # on blocking IO AND does NOT make a compound read-then-write atomic (a Thread
  # can be pre-empted between the two ops). So a single +Mutex+ guards EVERY
  # mutating critical section here — +bind+, +put_entity+, +bind_cas+, +unbind+
  # — making each one indivisible. This is required for correctness, not
  # decoration (profile [async].store_safety = mutex-guarded). Read-only
  # accessors also lock so they observe a consistent snapshot.
  #
  # == Emit pathway (§6.10 / v7.74 §6.13(c)) — the Core Extensibility Boundary
  #
  # Tree/content writes produce events delivered to registered consumers. The
  # hook is LIVE even with ZERO consumers (events are produced and discarded) so
  # a future extension can register a consumer WITHOUT rebuilding the peer — the
  # §6.13(c) MUST. Consumers fire OUTSIDE the lock (after the write commits) so a
  # consumer cannot deadlock on a re-entrant store call.
  class Store
    # A tree-change event (§6.10 Bind step).
    TreeEvent = Data.define(:event_type, :path, :new_hash, :previous_hash)
    # A content-store event (§6.10 Store step).
    ContentEvent = Data.define(:hash, :entity)

    def initialize
      @content = {}                 # hash-hex → Entity
      @tree = {}                    # path → hash-hex
      @content_consumers = []
      @tree_consumers = []
      @mutex = Mutex.new
    end

    # ── emit consumer registration (§6.10 consumer-registration primitive) ─────
    # Reachable any time, including post-bootstrap. Delivery is sync-inline (§9.4).

    def register_content_consumer(&block)
      @mutex.synchronize { @content_consumers << block }
    end

    def register_tree_consumer(&block)
      @mutex.synchronize { @tree_consumers << block }
    end

    # ── content store (§6.10 Store step: event fires only when entity is new) ──

    def put_entity(entity)
      event = nil
      @mutex.synchronize do
        k = hex(entity.content_hash)
        unless @content.key?(k)
          @content[k] = entity
          event = ContentEvent.new(hash: entity.content_hash, entity: entity)
        end
      end
      emit_content(event) if event
      nil
    end

    def get_by_hash(hash)
      @mutex.synchronize { @content[hex(hash)] }
    end

    # ── entity tree (§6.10 Bind step: event fires when the binding changes) ────

    def bind(path, entity)
      content_event, tree_event = @mutex.synchronize { commit_bind(path, entity) }
      emit_content(content_event) if content_event
      emit_tree(tree_event) if tree_event
      nil
    end

    # §3.9 compare-and-swap bind. +expected+ is the 33-byte hash the caller
    # believes is currently bound (a zero/empty hash means "expected absent").
    # The compare AND the swap happen in ONE critical section — the GVL alone
    # does NOT make this atomic (another thread can bind between a separate
    # read and write). Returns true on success, false on a CAS miss.
    def bind_cas(path, entity, expected)
      event = nil
      ok =
        @mutex.synchronize do
          current = @tree[path]
          matches =
            if expected.nil?
              true
            elsif zero_hash?(expected)
              current.nil?
            else
              current == hex(expected)
            end
          if matches
            event = commit_bind(path, entity)
            true
          else
            false
          end
        end
      if event
        content_event, tree_event = event
        emit_content(content_event) if content_event
        emit_tree(tree_event) if tree_event
      end
      ok
    end

    def unbind(path)
      event = nil
      @mutex.synchronize do
        prev = @tree.delete(path)
        event = TreeEvent.new(event_type: "deleted", path: path, new_hash: nil, previous_hash: prev) if prev
      end
      emit_tree(event) if event
      nil
    end

    # The hex content_hash bound at +path+, or nil.
    def hash_at(path)
      @mutex.synchronize { @tree[path] }
    end

    def get_at(path)
      @mutex.synchronize do
        h = @tree[path]
        h && @content[h]
      end
    end

    # One-level listing entry: a segment, its bound hash (or nil), and whether
    # the segment has deeper descendants.
    ListEntry = Data.define(:segment, :hash_hex, :has_children)

    # One-level listing under +prefix+ (a path; a trailing slash is added if
    # absent). Returns entries sorted by segment (§3.9).
    def listing(prefix)
      p = prefix.end_with?("/") ? prefix : "#{prefix}/"
      acc = {}                       # segment → [hash_or_nil, deeper]
      @mutex.synchronize do
        @tree.each do |path, hash|
          next unless path.length > p.length && path.start_with?(p)

          rest = path[p.length..]
          slash = rest.index("/")
          if slash
            seg = rest[0...slash]
            cell = (acc[seg] ||= [nil, false])
            cell[1] = true
          else
            cell = (acc[rest] ||= [nil, false])
            cell[0] = hash
          end
        end
      end
      acc.keys.sort.map { |seg| ListEntry.new(segment: seg, hash_hex: acc[seg][0], has_children: acc[seg][1]) }
    end

    private

    # MUST be called with @mutex held. Inserts the entity into the content store
    # and binds the path. Returns +[content_event_or_nil, tree_event_or_nil]+ for
    # the caller to emit AFTER releasing the lock (so a consumer cannot deadlock
    # on a re-entrant store call).
    def commit_bind(path, entity)
      k = hex(entity.content_hash)
      content_new = !@content.key?(k)
      @content[k] = entity if content_new
      content_event = content_new ? ContentEvent.new(hash: entity.content_hash, entity: entity) : nil
      prev = @tree[path]
      @tree[path] = k
      tree_event =
        if k == prev
          nil
        else
          TreeEvent.new(event_type: derive_event_type(prev, k), path: path, new_hash: k, previous_hash: prev)
        end
      [content_event, tree_event]
    end

    def derive_event_type(prev, nxt)
      return "created" if prev.nil?
      return "deleted" if nxt.nil?

      "modified"
    end

    def emit_content(event)
      @content_consumers.each { |c| c.call(event) }
    end

    def emit_tree(event)
      @tree_consumers.each { |c| c.call(event) }
    end

    def hex(bytes)
      bytes.b.unpack1("H*")
    end

    def zero_hash?(bytes)
      bytes.each_byte.all?(&:zero?)
    end
  end
end
